### PR TITLE
Explore raw data and initial testing of VAE model

### PR DESCRIPTION
**Pull request by ajlee21**
This PR attempts to train our existing VAE model on microbiome data, which is very sparse. Here I wanted to see if training would be possible on our existing model before trying to modify it to use a different loss function that will account for the sparsity in count data - https://www.nature.com/articles/s41467-018-07931-2

The main changes in this PR include:

1. Adding `0_explore_data.ipynb` notebook to do a first pass look of the raw data
2. Adding `1_train_vae.ipynb` notebook to train current VAE model on the raw data, but this will likely need to be modified given the sparsity of the data seen in the exploration notebook.

Some of the different things I've tried can be found [here](https://docs.google.com/presentation/d/1q1ZsX3wtKcwkwqHO3NlpTEMcXnUSFWKoNxyJ9FOohUY/edit?usp=sharing)

Overall, it looks like training the existing does a fair job training on normalized microbiome data. At this point, I think I want to evaluate what the VAE is learning/capturing - its very possible that its not learning anything

image

Feel free to let me know if you have any thoughts/suggestions on different params to try for training or ideas for evaluation

### CODE REVIEW

File: cm_modules/paths.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+"""Path definitions"""
+from pathlib import Path
+
+# Path to this repository
+PROJECT_DIR = Path(__file__).parents[1]
+
+# Path to local directory where data files will be stored
+LOCAL_DIR = Path.home()
+LOCAL_DATA_DIR = LOCAL_DIR / "Documents" / "Data" / "Common_microbes"
```
Comment (arielah) [inline cm_modules/paths.py:9]: Are your data and code being stored in the same directory? I've never used the pathlib package, so this may be ignorant of me, but I'm not sure why you need a PROJECT_DIR and a LOCAL_DIR

File: test_vae_training/0_explore_data.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
+# ---
```
Comment (arielah) [inline test_vae_training/0_explore_data.py:1]: FYI, the .ipynb version of this file isn't rendering in git. Not sure if it's a formatting issue or just git having problems. If it renders on your end there's probably no problem.

File: test_vae_training/0_explore_data.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import matplotlib.pyplot as plt
+from cm_modules import paths
+
+# +
+# Load dataset
+# Each column is a bacterial taxon, and each row is a single sample.
+# Rows are named using the format "${PROJECT_ID}_${SAMPLE_ID}"
+# Each cell is a read count.
+# There are ~11,800 samples in the preliminary dataset.
+# sample x taxon
```
Comment (arielah) [inline test_vae_training/0_explore_data.py]: Super nitpicky, but is the last line of the comment supposed to be there? It seems to say what you've already said.

File: test_vae_training/0_explore_data.py (hunk 2 above)
Comment (arielah) [reply test_vae_training/0_explore_data.py:1]: Nevermind, it's working now, definitely a git issue

File: test_vae_training/0_explore_data.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+microbiome_data_filename = paths.RAW_MICROBIOME_DATA
+microbiome_data = pd.read_csv(microbiome_data_filename, sep="\t", index_col=0, header=0)
+# -
+
+print(microbiome_data.shape)
+microbiome_data.head()
+
+# How sparse is this data matrix?
+# How many 0s across the entire matrix?
+num_zeros = (microbiome_data == 0).sum().sum()
```
Comment (arielah) [inline test_vae_training/0_explore_data.py:39]: You probably don't need to worry about this with your dataset size, but have you considered a sparse matrix format? Like single cell studies uses the `.mtx` file format, where instead of having a full matrix you list out the row and column ids and values for all nonzero cells. Could save space if your data is sparse enough.

File: test_vae_training/1_train_vae.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+# %load_ext autoreload
+# %autoreload 2
+# %matplotlib inline
+import os
+import matplotlib.pyplot as plt
+import pandas as pd
+from ponyo import utils, train_vae_modules
+
+# Set seeds to get reproducible VAE trained models
+train_vae_modules.set_all_seeds()
```
Comment (arielah) [inline test_vae_training/1_train_vae.py:29]: How much of a pain is it to switch to the new function calls being shown in the ipynb warnings? Could make the document a little easier to read.

File: test_vae_training/1_train_vae.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+params = utils.read_config(config_filename)
+
+dataset_name = params["dataset_name"]
+
+raw_compendium_filename = params["raw_compendium_filename"]
+normalized_compendium_filename = params["normalized_compendium_filename"]
+# -
+
+# Try normalzing the data
+train_vae_modules.normalize_expression_data(
```
Comment (arielah) [inline test_vae_training/1_train_vae.py:50]: How is this being normalized? My guess is transcripts per million or percentage of total sample reads, but it would be helpful to state explicitly in the comment.

File: test_vae_training/config_current_vae.tsv [hunk 7]
```diff
@@ -0,0 +1,14 @@
+local_dir	"/home/alexandra/Documents/Data/Common_microbes/"
```
Comment (arielah) [inline test_vae_training/config_current_vae.tsv:1]: I'm confused. It seems like you're using the pathlib script for part 0 and this config file for part 1? Why not just do one for both?

**Review by arielah — APPROVED** [review]
Looks good, I just had a couple of clarification questions. Interested to see where this analysis goes!

File: test_vae_training/0_explore_data.py (hunk 4 above)
Comment (ajlee21) [reply test_vae_training/0_explore_data.py:39]: Super helpful comment! I was just using the data format that was provided to me. Since I'm used to working in matrices and the size isn't an issue I think I'll leave it for now. But very helpful to know

File: test_vae_training/config_current_vae.tsv (hunk 7 above)
Comment (ajlee21) [reply test_vae_training/config_current_vae.tsv:1]: Good question. Unfortunately the reason there is both is because the ponyo library uses this config file whereas I've been trying to use pathlib for my new repositories. So its a little messy until I can update ponyo

File: test_vae_training/1_train_vae.py (hunk 6 above)
Comment (ajlee21) [reply test_vae_training/1_train_vae.py:50]: Here I am just 0-1 normalizing each taxon. I will add a comment here.

File: test_vae_training/1_train_vae.py (hunk 5 above)
Comment (ajlee21) [reply test_vae_training/1_train_vae.py:29]: So unfortunately pretty painful:) So `tf.compat.v1.ConfigProto` was introduced in TF v2.0 and currently ponyo is running on ponyo v <2.0 https://stackoverflow.com/questions/56127592/attributeerror-module-tensorflow-has-no-attribute-configproto

I have tried playing around trying to update TF for security reasons but I wasn't able to update it to v2.0 as other dependencies broke. 

I plan to update ponyo to use pytorch later so hopefully this will not be a headache for anyone trying to use ponyo and modify it later.

File: cm_modules/paths.py (hunk 1 above)
Comment (ajlee21) [reply cm_modules/paths.py:9]: Ah, so the PROJECT_DIR is pointing to the root directory of the repository where the code is stored but the LOCAL_DIR is pointing to the directory on my local machine where the data is being stored.

**Review by ben-heil — APPROVED** [review]
This project sounds really cool! I look forward to hearing about it next week. Back in the days of my youth I worked with some data from the American Gut Microbiome Project. I remember that one of the big questions on data quality was the existence of bacterial blooms. Since they shipped their samples at room temperature to sequencing facilities, the composition of the samples changed in transit. My understanding is that they handle these blooms by filtering out 
bacteria whose population increases at room temperature (the paper of how that actually works is [here](https://journals.asm.org/doi/full/10.1128/mSystems.00199-16). If you're looking for a potential application for a microbiome VAE, maybe you could calculate a vector that shifts between the bloom sample and original sample latent spaces and see if it outperforms their standard filtering method. No idea if that would be a good project, or something that you are interested in, but it may be worth looking into :shrug:

A few thoughts:

1. Are your taxa all at the same resolution? I know some microbiome software will output counts for the higher orders as well as their constituent parts. For example, you might have a Lactobacillaceae family feature as well as the individual lactobacillus species.   
2. Minor grammatical point that you might already be aware of: the singular form of taxa is taxon
3. It's weird that your val loss for the unnormalized data is spiking like that. 
4. I wrote a minor change to ponyo that displays the KL and reconstruction losses separately during training. You may already have code that does something similar, but if it'd be useful to you I can put together a PR

File: test_vae_training/0_explore_data.py (hunk 2 above)
Comment (ben-heil) [reply test_vae_training/0_explore_data.py:1]: In my experience, using https://nbviewer.jupyter.org/ has been easier than seeing whether github will decide to render a notebook

**Comment by ajlee21** [comment]
> This project sounds really cool! I look forward to hearing about it next week. Back in the days of my youth I worked with some data from the American Gut Microbiome Project. I remember that one of the big questions on data quality was the existence of bacterial blooms. Since they shipped their samples at room temperature to sequencing facilities, the composition of the samples changed in transit. My understanding is that they handle these blooms by filtering out
> bacteria whose population increases at room temperature (the paper of how that actually works is [here](https://journals.asm.org/doi/full/10.1128/mSystems.00199-16). If you're looking for a potential application for a microbiome VAE, maybe you could calculate a vector that shifts between the bloom sample and original sample latent spaces and see if it outperforms their standard filtering method. No idea if that would be a good project, or something that you are interested in, but it may be worth looking into 

This is helpful thanks Ben! So for now I think my goal is just to see if I can successfully get a VAE to train on this type of data. The current application that was discussed was to be an extension of SOPHIE, so trying to filter for commonly expressed microbes instead of genes. That project sounds interesting though and something I'll make a note of

> A few thoughts:
> 
> 1. Are your taxa all at the same resolution? I know some microbiome software will output counts for the higher orders as well as their constituent parts. For example, you might have a Lactobacillaceae family feature as well as the individual lactobacillus species.

Good question. I didn't look into that nor did I ask the data provider. I will need to take a look at this in the future

> 2. Minor grammatical point that you might already be aware of: the singular form of taxa is taxon

Thanks for the catch

> 3. It's weird that your val loss for the unnormalized data is spiking like that.

Yes, I'm not sure how to explain this finding either. Something I'll need to think about, especially if I need to modify the loss to use counts.  

> 4. I wrote a minor change to ponyo that displays the KL and reconstruction losses separately during training. You may already have code that does something similar, but if it'd be useful to you I can put together a PR

I'm trying to decide if users will want to see the plots separately. The reason they would is if they are trying to debug the training, which may happen depending on their application. So yes please make a PR or you can make an issue and I'll add it later, I don't think there is a big rush on this feature.
